packages/apps-engine/src/server/accessors/Http.ts

```diff
@@ -31,47 +37,76 @@ export class Http implements IHttp {
         return this._processHandler(url, RequestMethod.PATCH, options);
     }
 
-    private async _processHandler(url: string, method: RequestMethod, options?: IHttpRequest): Promise<IHttpResponse> {
-        let request = options || {};
-
-        if (typeof request.headers === 'undefined') {
-            request.headers = {};
-        }
-
-        this.httpExtender.getDefaultHeaders().forEach((value: string, key: string) => {
-            if (typeof request.headers[key] !== 'string') {
-                request.headers[key] = value;
+    private async _processHandler(
+        url: string,
+        method: RequestMethod,
+        options?: IHttpRequest
+    ): Promise<IHttpResponse> {
+        const retryConfig = {
+            ...Http.DEFAULT_RETRY_CONFIG,
+            ...options?.retry
+        };
+
+        let attempt = 0;
+        let lastError: any;
+
+        while (attempt < (retryConfig.maxAttempts)) {
+            try {
+                let request = options || {};
+
+                if (typeof request.headers === 'undefined') {
+                    request.headers = {};
+                }
+
+                this.httpExtender.getDefaultHeaders().forEach((value: string, key: string) => {
+                    if (typeof request.headers[key] !== 'string') {
+                        request.headers[key] = value;
+                    }
+                });
+
+                if (typeof request.params === 'undefined') {
+                    request.params = {};
+                }
+
+                this.httpExtender.getDefaultParams().forEach((value: string, key: string) => {
+                    if (typeof request.params[key] !== 'string') {
+                        request.params[key] = value;
+                    }
+                });
+
+                const reader = this.accessManager.getReader(this.appId);
+                const persis = this.accessManager.getPersistence(this.appId);
+
+                for (const handler of this.httpExtender.getPreRequestHandlers()) {
+                    request = await handler.executePreHttpRequest(url, request, reader, persis);
+                }
+
+                let response = await this.bridges.getHttpBridge().doCall({
+                    appId: this.appId,
+                    method,
+                    url,
+                    request,
+                });
+
+                for (const handler of this.httpExtender.getPreResponseHandlers()) {
+                    response = await handler.executePreHttpResponse(response, reader, persis);
+                }
+
+                return response;
+
+            } catch (error) {
+                lastError = error;
+                if (!retryConfig.enabled || attempt === (retryConfig.maxAttempts || 3) - 1) {
+                    throw error;
+                }
+
+                // Exponential backoff
+                const delay = retryConfig.initialDelay * Math.pow(2, attempt);
```

Comment:
I think we should probably be smarter here instead of doing exponential backoff despite the return of the request.
For example, let's say the server returned a 404. In that case, it would be pointless to keep requesting the server for the same resource.

I think we should probably just re-do the calls for service unavailable status code (503) and (maybe) 500, not sure though, otherwise, I think re-doing the call is detrimental, and failure logic should be handled by the apps themselves WDYT?